### PR TITLE
Fix: tdr_anvil's fetch_bundle lacks test coverage (#5046, #5898)

### LINKED ISSUE #5046: tdr_anvil's fetch_bundle lacks test coverage
…due to TinyQuery not supporting the `WITH` clause

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [ ] … ~introduce a new software dependency like https://github.com/DataBiosphere/azul/issues/5046#issuecomment-1915746834
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

### LINKED ISSUE #5898: AnVIL plugin skips donor.part_of_dataset_id validation
Due to a pluralization mismatch, [this if statement](https://github.com/DataBiosphere/azul/blob/eb3bb69d6f2788bb2030705230ae38ee878b3c0f/src/azul/plugins/repository/tdr_anvil/__init__.py#L331) is never executed.

The most immediate consequence of this is that we skip the validation that `biosample.part_of_dataset_id` and `donor.part_of_dataset_id` are redundant. However it also has the side effect of causing the extraneous donor field to persist and be passed to the metadata plugin. This has no effect on the service, but I expect that the field is passed through to the replicas index. Therefore, we would need to reindex to fully remove all traces all of this bug. However, since we currently don't read from the replicas index, we could postpone the reindex.

The new test coverage provided in the resolution to https://github.com/DataBiosphere/azul/issues/5046 is what caught this bug.

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #5046, #5898

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>
- [x] All connected issues are resolved partially <sub>or this PR does not have the `partial` label</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading deployments)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading deployments</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from system administrator
- [x] PR is assigned to system administrator

### System administrator (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvilprod`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only include `p` if the PR is labeled `partial`</sub>
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvilprod`</sub>
- [x] Considered deindexing individual sources in `dev` <sub>or this PR does not merely remove sources from existing catalogs in `dev`</sub>
- [x] Considered deindexing individual sources in `anvildev` <sub>or this PR does not merely remove sources from existing catalogs in `anvildev`</sub>
- [x] Considered deindexing individual sources in `anvilprod` <sub>or this PR does not merely remove sources from existing catalogs in `anvilprod`</sub>
- [x] Considered indexing individual sources in `dev` <sub>or this PR does not merely add sources to existing catalogs in `dev`</sub>
- [x] Considered indexing individual sources in `anvildev` <sub>or this PR does not merely add sources to existing catalogs in `anvildev`</sub>
- [x] Considered indexing individual sources in `anvilprod` <sub>or this PR does not merely add sources to existing catalogs in `anvilprod`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing `anvilprod`</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing `anvilprod`</sub>

### Operator

- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by achave11-ucsc — DISMISSED** [review]

File: test/indexer/test_anvil.py [hunk 1]
```diff
@@ -109,6 +119,17 @@ def test_indexing(self):
...
                         self.index_service.delete_indices(self.catalog)
 
+    def test_fetch_bundle(self):
+        canned_bundle = self._load_canned_bundle(self.bundle)
+        assert isinstance(canned_bundle, TDRAnvilBundle)
+        self._make_mock_tdr_tables(self.bundle)
+        plugin = self.plugin_for_source_spec(canned_bundle.fqid.source.spec)
+        bundle = plugin.fetch_bundle(self.bundle)
+        assert isinstance(bundle, TDRAnvilBundle)
+        self.assertEqual(canned_bundle.fqid, bundle.fqid)
```
Comment (hannes-ucsc) [inline test/indexer/test_anvil.py:129]: These are fairly weak assertions. Can we compare the result to the "intermediate" can?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/test_anvil.py (hunk 1 above)
Comment (hannes-ucsc) [reply test/indexer/test_anvil.py:129]: Please disregard.

File: test/indexer/test_tdr.py [hunk 2]
```diff
@@ -274,27 +288,31 @@ def test_fetch_bundle(self):
         bundle = self._load_canned_bundle(self.bundle_fqid)
         # Test valid links
         self._test_fetch_bundle(bundle, load_tables=True)
-
-        # FIXME: Rewrite this part to use BigQuery DML statements
-        #        https://github.com/DataBiosphere/azul/issues/5046
-        if False:
-            self._test_invalid_links(bundle)
+        # Test invalid links
+        self._test_invalid_links(bundle)
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py]: Given the method name, the comment is redundant. Feel free to inline the method. I just extracted the it to minimize the diff that added the FIXME.

File: test/indexer/test_tdr.py [hunk 3]
```diff
@@ -305,7 +323,11 @@ def _test_invalid_links(self, bundle: TDRHCABundle):
...
         # Update table
-        links_content_column[0] = json.dumps(links_content)
+        plugin.tdr.run_sql(f'''
+            UPDATE {plugin._full_table_name(spec, 'links')}
+            SET content = {json.dumps(links_content)!r}
+            WHERE links_id = "{links_id}"
+        ''')
         # Invoke code under test
         with self.assertRaises(RequirementError):
             self._test_fetch_bundle(bundle, load_tables=False)
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py]: This appears to be the same code as above, just with a different key and value. Try roll this into a loop.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/repository/tdr_anvil/__init__.py [hunk 4]
```diff
@@ -328,12 +328,13 @@ def _primary_bundle(self, bundle_fqid: TDRAnvilBundleFQID) -> TDRAnvilBundle:
...
                 # We expect that the foreign key `part_of_dataset_id` is
                 # redundant for biosamples and donors. To simplify our queries,
                 # we do not follow the latter during the graph traversal.
-                # Here, we validate our expectation.
-                dataset_id: Key = one(keys_by_type['datasets'])
+                # Here, we validate our expectation. Note that the key is an
+                # array for biosamples, but not for donors.
+                dataset_id: Key = one(keys_by_type['dataset'])
                 for row in rows:
                     require(row.pop('part_of_dataset_id') == dataset_id)
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr_anvil/__init__.py]: The commit title for this hunk surprises me. The body of the `if` has a side effect that is visible beyond the `if` so I wonder why this described as just a broken requirement.

Comment (nadove-ucsc) [reply src/azul/plugins/repository/tdr_anvil/__init__.py]: I had mistakenly assumed that the extraneous presence of the `part_of_dataset_id` field would not affect the behavior of the indexer in any way, but now I realize that this field will now disappear from the contents of donors in the `replica` index. Would that mean that this PR now requires a reindex?

Comment (nadove-ucsc) [reply src/azul/plugins/repository/tdr_anvil/__init__.py]: Or do we need a separate ticket/PR to cover this bug?

Comment (hannes-ucsc) [reply src/azul/plugins/repository/tdr_anvil/__init__.py]: Let's factor this into a separate ticket, if possible.

Comment (nadove-ucsc) [reply src/azul/plugins/repository/tdr_anvil/__init__.py]: https://github.com/DataBiosphere/azul/issues/5898

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by dsotirho-ucsc** [comment]
PR failed IT on hammerbox:

https://gitlab.prod.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/20974

```
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest.test_indexing)
----------------------------------------------------------------------
Traceback (most recent call last):
…
  File "/builds/ucsc/azul/src/azul/queues.py", line 270, in wait_to_stabilize
    raise Exception('The queues have stalled', final_length)
Exception: ('The queues have stalled', 7)
----------------------------------------------------------------------
```

Hammebox fail queue: 
[azul-notifications_fail-hammerbox.json](https://github.com/DataBiosphere/azul/files/14101653/azul-notifications_fail-hammerbox.json)


[CloudWatch logs](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2024-01-30T09*3a39*3a00.000Z~start~'2024-01-30T08*3a53*3a41.428Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40message*20like*20*2fERROR*2f*0a*7c*20sort*20*40timestamp*20desc*0a*7c*20limit*20200~queryId~'c5bcdb3580a85712-403f2238-4f76bcd-9e8daa5-8259a5388a9222a2a7a7a~source~(~'*2faws*2flambda*2fazul-indexer-hammerbox-aggregate~'*2faws*2flambda*2fazul-indexer-hammerbox-contribute~'*2faws*2flambda*2fazul-indexer-hammerbox-contribute_retry~'*2faws*2flambda*2fazul-indexer-hammerbox-indexercachehealth~'*2faws*2flambda*2fazul-indexer-hammerbox))):

```
[ERROR] RequirementError
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 166, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1756, in __call__
    return self.handler(event_obj)
  File "/var/task/app.py", line 337, in contribute_retry
    app.index_controller.contribute(event, retry=True)
  File "/var/task/azul/indexer/index_controller.py", line 180, in contribute
    contributions, replicas = self.transform(catalog, notification, delete)
  File "/var/task/azul/indexer/index_controller.py", line 228, in transform
    bundle = service.fetch_bundle(catalog, bundle_fqid)
  File "/var/task/azul/indexer/index_service.py", line 193, in fetch_bundle
    return plugin.fetch_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr.py", line 212, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 288, in _emulate_bundle
    return self._primary_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 339, in _primary_bundle
    require(row.pop('part_of_dataset_id') == dataset_id)
  File "/var/task/azul/__init__.py", line 1531, in require
    reject(not condition, *args, exception=exception)
  File "/var/task/azul/__init__.py", line 1550, in reject
    raise exception(*args)
```

**Comment by nadove-ucsc** [comment]
One of the failing snapshots [has already been removed](https://github.com/DataBiosphere/azul/commit/1b7466fbe02a25244b90896b809fa18c5d190aa6).

**Comment by nadove-ucsc** [comment]
The following snapshots on `hammerbox` contain donor rows that will be rejected:

 - ANVIL_CCDG_Baylor_CVD_EOCAD_SoL_WGS_20230418_ANV5_202310101651
 - ANVIL_CCDG_WashU_CVD_PAGE_HMB_NPU_WGS_20221025_ANV5_202304211801
 - ANVIL_CMG_BaylorHopkins_HMB_IRB_NPU_WES_20221020_ANV5_202304211516
 - ANVIL_CMG_BaylorHopkins_HMB_NPU_WES_20230525_ANV5_202306211834
 - ANVIL_CMG_Broad_Muscle_Beggs_WES_20221102_ANV5_202304241506
 - ANVIL_CMG_YALE_DS_MC_20221026_ANV5_202304211522
 - ANVIL_CMH_GAFK_SCATAC_20221107_ANV5_202304211531
 - ANVIL_CMH_GAFK_WGBS_20230327_ANV5_202304211534
 - ANVIL_NIMH_Broad_WGSPD1_McCarroll_Pato_GRU_WGS_20221115_ANV5_202304242056

In all cases, the rejection is caused by donor rows where the `part_of_dataset_id` foreign key is `NULL`.

We could remove the snapshots, but it's worth noting that it would be consistent with the intent behind the `reject` invocation to accept these rows as valid. The purpose of the `reject` is to verify that `anvil_donor.part_of_dataset_id` does not contain different information than `anvil_biosample.part_of_dataset_id`. Since `NULL` indicates an absence of information, we could rationalize that there is no contradiction between these fields when one of them is `NULL`.

**Comment by nadove-ucsc** [comment]
I also checked `anvildev` and found no problematic donors.

**Comment by nadove-ucsc** [comment]
Author to confirm whether any biosample rows in any snapshots contain one of the following values:

1. NULL 
2. Empty array
3. Array containing NULL

**Comment by nadove-ucsc** [comment]
Decided at PL to proceed with removing the defective snapshots listed above.

**Comment by nadove-ucsc** [comment]
> Author to confirm whether any biosample rows in any snapshots contain one of the following values:

None found in `anvilbox` or `hammerbox` (including in the removed snapshots).

Worth noting that I also found no biosamples where the array has a size other than 1.

**Comment by hannes-ucsc** [comment]
I notified BI about this: https://ucsc-gi.slack.com/archives/C03TPJS54DC/p1706906651968669

**Review by hannes-ucsc — APPROVED** [review]

**Comment by dsotirho-ucsc** [comment]
The 9 removed snapshots were manually deindexed from the `anvil3` and `anvil4` catalogs using the `reindex.py` script for both `hammerbox` and `anvilprod`, rather than doing a full reindex.

```
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-6ddb7e8d:snapshot/ANVIL_CMG_BaylorHopkins_HMB_IRB_NPU_WES_20221020_ANV5_202304211516*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-d09f6f5a:snapshot/ANVIL_CMG_BaylorHopkins_HMB_NPU_WES_20230525_ANV5_202306211834*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-937b5d92:snapshot/ANVIL_CMG_Broad_Muscle_Beggs_WES_20221102_ANV5_202304241506*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-19b85efe:snapshot/ANVIL_CMG_YALE_DS_MC_20221026_ANV5_202304211522*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-f2af4233:snapshot/ANVIL_CCDG_WashU_CVD_PAGE_HMB_NPU_WGS_20221025_ANV5_202304211801*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-3b296131:snapshot/ANVIL_CMH_GAFK_SCATAC_20221107_ANV5_202304211531*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-8acbf07f:snapshot/ANVIL_CMH_GAFK_WGBS_20230327_ANV5_202304211534*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-7c00e8e5:snapshot/ANVIL_NIMH_Broad_WGSPD1_McCarroll_Pato_GRU_WGS_20221115_ANV5_202304242056*
python scripts/reindex.py --catalogs anvil3 anvil4 --deindex --sources tdr:datarepo-a749913a:snapshot/ANVIL_CCDG_Baylor_CVD_EOCAD_SoL_WGS_20230418_ANV5_202310101651*
```

**Pull request merged**